### PR TITLE
Refactor to allow for @NamespaceAccessible

### PR DESCRIPTION
**Pull request by vazexqi**
We are using this library for a project that we are doing at Salesforce. For our use case, we need to use it in a 2GP setting. This means that this is in something like a base package and we have other packages that reference it.

The library as-is doesn't work for 2GP since it doesn't export its public constructs via [`@NamespaceAccessible`](https://developer.salesforce.com/docs/atlas.en-us.apexcode.meta/apexcode/apex_classes_annotation_NamespaceAccessible.htm)

This PR refactors the current version of the library to support that.

Full disclosure: I am gradually using parts of the library so I am not an expert on all parts of it. I have relied on the tests during my refactoring.

My methodology was to create two packages: parent and child. I left the fflib-apex-mocks non-test classes in parent. And I moved the tests to child. I then made my changes to add `@NamespaceAccessible`. I package both packages and then ran the tests. My thought process is that this will help verify any major issues.

I then moved all the changes back into this non-namespaced package and ran the tests in my non-namespaced scratch org.

High-level summary of changes. I've tried to keep formatting changes to a minimum.

1. Introduce `@NamespaceAccessible` to all public classes, methods, etc. I needed to bump the class version to 43.0 or higher since that was when that annotation was supported.
1. Due to https://success.salesforce.com/issues_view?Id=a1p3A0000003gpX, I had to refactor some properties to use getters/setters as appropriate.
1. Due to https://github.com/JonnyPower/second-gen-packaging-public-interface-implentation#issue-with-2gp-and-interfaces, I had to refactor `fflib_IMatcher` to be an abstract class.
1. Move `fflib_Inheritor.cls` to test since it felt odd to be in the main folder.

The above changes may or may not be palatable to the maintainers of this library. I have raised the two bugs up with Chris Peterson, Apex PM, and he knows of them but has not definitive timeline yet to fix them.

### CODE REVIEW

**Comment by phardaker** [comment]
Great to hear you're using ApexMocks internally @vazexqi :-)

**Comment by afawcett** [comment]
@vazexqi thanks for raising these. It seems though that merging this would break those already using the library or no? For example moving from an interface to an abstract base class?

**Comment by vazexqi** [comment]
Yes, you are right – it will break them if they just replace their current version with this new one. However, my understanding of how this library is used is that a user has to *explicitly* pull down the latest version of this library and put it into the source code. So, the user needs to decide to incorporate the latest version of the library. 

I'm not as familiar with the upgrade path/policies for this library and maybe that path is not acceptable.

Ideally, interfaces would be supported in `@NamespaceAccessible`. I've spoken to Chris Peterson about getting `@NamespaceAccessible` to support interfaces. It will be, [insert safe harbor] at least one more release out. So anyone looking to take advantage of this library in a 2GP setting across multiple packages is also at an impasse without these changes.

It's a hard choice. And, the correct decision might be to *not* merge things in but to keep this in _separate_ branch for those who do want to use this in 2GP _now_ and can tolerate this change from interface to abstract classes. 

I'm open to either.

**Comment by afawcett** [comment]
Thanks @vazexqi  for confirming my understanding. I would like thoughts from other contributors on this adoption inhibitor. Your assertion on usage is in my experience partially the case, in other cases, folks use these things in ISV packages or just upgrade by deploying over a previous deployment.

**Comment by daveespo** [comment]
Using the trick from #268, I ran a search for 'implements fflib_IMatcher" (https://github.com/search?q=implements+fflib_IMatcher+-filename%3Afflib_*++-extension%3Ahtml&type=Code&ref=advsearch&l=&l=) and turned up 4 results. That obviously doesn't confirm that no one is implementing their own matchers, but it suggests that it's infrequent... so perhaps breaking backward compat for `fflib_IMatcher` isn't a big concern

As an ISV, I have trouble seeing the utility of `NamespaceAccessible` until Bundles are supported (i.e. multiple package installation in a single atomic operation) since orchestrating multi-package installations is a huge drag. I don't believe Bundles are due until Winter/Spring '21. That's another way of saying that code reuse across multiple packages published by the same vendor is still hobbled with problems and duplicating code is still generally the least problematic of a few bad choices. 

I'm ambivalent on merging these changes -- For those that are doing multi-package projects, I can see why you'd like this merged in.

@ImJohnMDaniel  .. thoughts?

File: sfdx-source/apex-mocks/main/classes/fflib_InOrder.cls-meta.xml [hunk 1]
```diff
@@ -1,5 +1,5 @@
-<?xml version="1.0" encoding="UTF-8"?>
+<?xml version="1.0" encoding="UTF-8" ?>
 <ApexClass xmlns="http://soap.sforce.com/2006/04/metadata">
-    <apiVersion>30.0</apiVersion>
+    <apiVersion>43.0</apiVersion>
```
Comment (daveespo) [inline sfdx-source/apex-mocks/main/classes/fflib_InOrder.cls-meta.xml:3]: s/b API 48

**Comment by vazexqi** [comment]
I've given the comments some thought here.

Thinking ahead to where we would like this to be when everything on the Salesforce side is ready would be to keep the interfaces but annotate them with `@NamespaceAccessible`, not change them into abstract classes. Also, we would likely want to use proper `get;set` and not convert things into setters. Thus this PR represents a potential churn.

I would like to propose the following:

1. We put this into its own branch (so that we don't pollute the history).
1. We cut a formal GitHub Release from this branch and tag it with version 43.0
1. We advertise this on Twitter etc and _then_ we track the download of the release using https://github.community/t5/How-to-use-Git-and-GitHub/Counting-number-of-downloads-of-a-release/td-p/1175 Admins, you should have access to these numbers?

If the numbers show that developers do use `@NamespaceAccessible` for 2GP multi-package development even in their current form, then we can consider merging this into master.
If the numbers show that developers do not use it, then we can leave this in a branch.

Thoughts?

**Comment by afawcett** [comment]
@daveespo love the detective work!
@vazexqi  love the data driven approach!

**Comment by daveespo** [comment]
@vazexqi  -- I don't think it's realistic to create a 'forked' version of this project and keep it maintained over time. The nature of the changes on this branch would make pulling in upstream changes problematic (merge conflicts) and would require continuous attention to applying the NamespaceAccessible annotation to any new methods added in the upstream branch as they're merged in.

Nothing is stopping people from pulling this branch directly and using it in their project .. so perhaps that's where we should leave this issue until proper platform support arrives for finishing the loose ends with NamespaceAccessible. I'm fine with putting it on a branch in this repo vs. your personal repo.

@ImJohnMDaniel  -- thoughts?

**Comment by ImJohnMDaniel** [comment]
@vazexqi -- Just circling back to this PR.  Do you have much in the way of stats yet?

**Comment by vazexqi** [comment]
@ImJohnMDaniel - Thanks for following up.

If you look at https://trailblazer.salesforce.com/issues_view?Id=a1p3A0000003gpX you would see that it's been scheduled for 'Winter 21'. Now, while this is not a guarantee, it is promising ;-)

This branch is going to be defunct when that goes out so we should *not* merge this in.

**Comment by ImJohnMDaniel** [comment]
understood.

Is there any reason to keep the PR open then?

**Comment by vazexqi** [comment]
No reason unless there's a party that's interested in using this before Winter 21 (safe harbor). I know we are using it internally but plan to switch as soon as that is available.

I think we should _close_ this PR.

Anyone who is interested in it will probably do a google search for the phrase "namespaceaccessible apex interface" and this will come up. And, I won't delete my branch yet so they can still use it if they wish.

**Comment by ImJohnMDaniel** [comment]
Sounds good.  Thanks for the feedback.  

@afawcett and @daveespo -- any objections to closing this PR?

**Comment by daveespo** [comment]
I concur; close without merge

**Comment by attilah** [comment]
Now we have Winter'21 released and it does support `@namespaceAccessible` on interfaces and more, I think this issue could be reopened ;-)

**Comment by ImJohnMDaniel** [comment]
Thanks for the suggestion @attilah.  I just opened #102 to look into this.

**Pull request closed without merging**